models/collectionswap/ethereum/collectionswap_ethereum_events.sql

```diff
@@ -0,0 +1,163 @@
+                ,'Buy' as trade_category
+                ,nftIds as nft_id_array
+                ,cardinality(nftIds) as number_of_items
+                ,cast(outputAmount as decimal(38)) as amount_raw
+                ,cast(protocolFee as decimal(38)) as platform_fee_amount_raw
+                ,get_json_object(royaltyDue[0], '$.amount') as royalty_fee_amount_raw
+                ,get_json_object(royaltyDue[0], '$.recipient') as royalty_fee_receive_address
+                ,cast(tradeFee as decimal(38)) as trade_fee_amount_raw
+                ,contract_address as project_contract_address
+            from collectionswap_ethereum.CollectionPool_evt_SwapNFTOutPool e
+            {% if is_incremental() %}
+            WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+            {% else %}
+            WHERE evt_block_time >= '{{project_start_date}}'
+            {% endif %}
+            union all
+            select
+                evt_block_number as block_number
+                ,evt_block_time as block_time
+                ,evt_index
+                ,evt_tx_hash as tx_hash
+                ,contract_address as buyer
+                ,null as seller
+                ,'Sell' as trade_category
+                ,nftIds as nft_id_array
+                ,cardinality(nftIds) as number_of_items
+                ,cast(inputAmount + protocolFee + cast(get_json_object(royaltyDue[0], '$.amount') as decimal(38)) as decimal(38)) as amount_raw
+                ,cast(protocolFee as decimal(38)) as platform_fee_amount_raw
+                ,get_json_object(royaltyDue[0], '$.amount') as royalty_fee_amount_raw
+                ,get_json_object(royaltyDue[0], '$.recipient') as royalty_fee_receive_address
+                ,cast(tradeFee as decimal(38)) as trade_fee_amount_raw
+                ,contract_address as project_contract_address
+            from collectionswap_ethereum.CollectionPool_evt_SwapNFTInPool e
+            {% if is_incremental() %}
+            WHERE evt_block_time >= date_trunc("day", now() - interval '1 week')
+            {% else %}
+            WHERE evt_block_time >= '{{project_start_date}}'
+            {% endif %}
+            )
+    )
+),
+
+base_trades as (
+    select
+    t.*,
+    p.nft_contract_address,
+    case when p.token_address = '0x0000000000000000000000000000000000000000'
+        then '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'
+        else p.token_address END as currency_contract
+    from raw_trades t
+    left join {{ ref('collectionswap_ethereum_pools') }} p
+    on t.project_contract_address = p.pool_address
+)
+
+
+
+select
+    'ethereum' as blockchain
+    ,'collectionswap' as project
+    ,'v1' as version
+    ,date_trunc('day',t.block_time ) as block_date
+    ,t.block_time
+    ,t.block_number
+    ,t.evt_index
+    ,t.tx_hash
+    ,t.project_contract_address
+    ,t.nft_contract_address
+    ,t.token_id
+    ,coalesce(t.buyer, tx.`from`) as buyer
```

Comment:
will `tx.from` always have a value, to ensure the `null` hardcoded above in CTE won't push through? we've had datatype issues with that

Reply:
Yes, it should have. I'll change the matching to an inner join which gives us the guarantee.